<a href="https://colab.research.google.com/github/rujapa/digitaladvertisingAPRD/blob/master/Python_exercise_3_Using_LASSO_Regression_to_Model_Fat_Sales_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Python Exercise 3: Using LASSO Regression to Model Fat Sales Data

In [0]:
#import library
import numpy as np
import pandas as pd
import seaborn as sns
import statsmodels.api as sm
import matplotlib.pyplot as plt
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all" 

In [0]:
#import csv file
from google.colab import drive
drive.mount('/content/drive')
alldata=pd.read_csv(open('drive/My Drive/APRD Digital ads/Homework/finalmaster-ratios.csv','rb'))

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LassoLarsCV
import matplotlib.pyplot as plt

In [0]:
#create a list of all the predictors that we want to feed in the LassoLarsCV model
allvariablenames = list(alldata.columns.values)

In [0]:
allvariablenames

['# Purchases',
 'B01001001',
 'B01001002',
 'B01001003',
 'B01001004',
 'B01001005',
 'B01001006',
 'B01001007',
 'B01001008',
 'B01001009',
 'B01001010',
 'B01001011',
 'B01001012',
 'B01001013',
 'B01001014',
 'B01001015',
 'B01001016',
 'B01001017',
 'B01001018',
 'B01001019',
 'B01001020',
 'B01001021',
 'B01001022',
 'B01001023',
 'B01001024',
 'B01001025',
 'B01001026',
 'B01001027',
 'B01001028',
 'B01001029',
 'B01001030',
 'B01001031',
 'B01001032',
 'B01001033',
 'B01001034',
 'B01001035',
 'B01001036',
 'B01001037',
 'B01001038',
 'B01001039',
 'B01001040',
 'B01001041',
 'B01001042',
 'B01001043',
 'B01001044',
 'B01001045',
 'B01001046',
 'B01001047',
 'B01001048',
 'B01001049',
 'B02001001',
 'B02001002',
 'B02001003',
 'B02001004',
 'B02001005',
 'B02001006',
 'B02001007',
 'B02001008',
 'B02001009',
 'B02001010',
 'B12001001',
 'B12001002',
 'B12001003',
 'B12001004',
 'B12001005',
 'B12001006',
 'B12001007',
 'B12001008',
 'B12001009',
 'B12001010',
 'B12001011',
 'B1

In [0]:
alldata

# Purchases  B01001001   B01001002  ...   B19001015  B19001016  B19001017
0             22     206252  469.226965  ...   36.440765  23.446284  21.197485
1              7      61399  486.538869  ...   33.000508  33.169741  24.792689
2              3      73170  489.859232  ...   20.125056  11.890525  16.537397
3             94     251724  505.585483  ...   43.731067  38.851729  40.427349
4              0      37382  495.586111  ...   36.118530  31.603714  19.648989
..           ...        ...         ...  ...         ...        ...        ...
727            4      38106  482.312497  ...   27.706620  19.347445  15.033032
728            3      32531  511.665796  ...  106.653459  84.530296  76.027737
729            3      38761  493.511519  ...   49.390327  47.814769  31.031648
730            1      28359  496.773511  ...   29.939545  24.085980  19.192016
731          507    5861000  484.835011  ...   40.966917  44.925312  52.695060

[732 rows x 190 columns]

In [0]:
listofallpredictors = allvariablenames[8:]

In [0]:
#load predictors into dataframe
predictors = alldata[listofallpredictors]  
#load target into dataframe
target = alldata['# Purchases']   

In [0]:
# split data into train and test sets, with 30% retained for test
pred_train, pred_test, tar_train, tar_test = train_test_split(predictors, target, test_size=.3, random_state=123)    

In [0]:
from sklearn.linear_model import LassoLarsCV
from sklearn.datasets import make_regression

In [0]:
X, y = make_regression(noise=4.0, random_state=0)
#model = LassoLarsCV(precompute=False,cv=10).fit(X, y)
#model.score(X, y)
model = LassoLarsCV(precompute=False, cv=10)
model.fit(pred_train, tar_train)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 5 iterations, i.e. alpha=1.496e+00, with an active set of 5 regressors, and the smallest cholesky pivot element being 6.322e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 8 iterations, i.e. alpha=1.098e+00, with an active set of 8 regressors, and the smallest cholesky pivot element being 6.322e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 10 iterations, i.e. alpha=7.329e-01, with an active set of 10 regressors, and the smallest cholesky pivot element b

LassoLarsCV(copy_X=True, cv=10, eps=2.220446049250313e-16, fit_intercept=True,
            max_iter=500, max_n_alphas=1000, n_jobs=None, normalize=True,
            positive=False, precompute=False, verbose=False)

In [0]:
#build coefficent chart

predictors_model=pd.DataFrame(listofallpredictors)
predictors_model.columns = ['label']
predictors_model['coeff'] = model.coef_

for index, row in predictors_model.iterrows():
    if row['coeff'] > 0:
        print(row.values)

['B01001014' 0.8558761066941788]
['B01001036' 2.5053482381631653]
['B01001037' 0.8892493223320962]
['B01001038' 1.5316387928880384]
['B02001005' 0.41252295298457853]
['B13014026' 0.48004105312075906]
['B13014027' 0.6978957445987839]
['B13016001' 875149895.329212]
['B19001017' 1.4834348068681533]


Question 1 : In your own words, explain what the above lines of code are doing. Why am I doing it? Explain each line.

Question 2 : Interpret each variable intuitively. What Census variables most predict sales? What does that mean, practically? Here's what I'd put for the example above. "In areas where there are more females aged 30-34, we sell more Bobo Bars."

Question 3 : If I had to report only two census variables to my boss that most steeply predicted sales, what would those be?

In [0]:
# mean squared error 
from sklearn.metrics import mean_squared_error
train_error = mean_squared_error(tar_train, model.predict(pred_train))
print ('training data MSE')
print(train_error)

test_error = mean_squared_error(tar_test, model.predict(pred_test))
print ('testing data MSE')
print(test_error)

training data MSE
22025.491066757
testing data MSE
41549.54803776253


Question 4 Are the training and text set mean squared errors similar? What does that mean practically? 

In [0]:
#r squared
rsquared_train=model.score(pred_train,tar_train)
print ('training data R-square')
print(rsquared_train)


training data R-square
0.2400221219784492


In [0]:
rsquared_test = model.score(pred_test, tar_test)
print ('testing data R-square')
print(rsquared_test)

testing data R-square
0.1758628512005107


The MSE for training data(22025.49) and testing data (41549.548) are significantly different. The training data MSE is fitting better than testing data becasue the MSE is a measure of the quality of an estimator it is always non-negative, and values closer to zero are better. 
Moreover, Compare the two R Squared values The R squared value for the training set is slightly higher, indicating that this model is slight more accurate at explaining the variability in the data

Question 5  If your boss asked, "How well does Census data, overall, predict sales?" What would you say? Why?

Finally, let's see what our y-intercept is, so we can interpret what our baseline sales number looks like, all things considered:

In [0]:
print("y interecept:")
print(model.intercept_)

y interecept:
22.19738813257551


Question 6 : What is our baseline sales number? What does that mean, practically? Think back to what y-intercepts mean in regression models.


Our baseline sales number (y-intercept) is 2.75. This means that if all other variables are zero, we sell an average of 2.75 bars per customer.